python/langsmith/evaluation/_runner.py

```diff
@@ -0,0 +1,895 @@
+    def __iter__(self) -> Iterator[ExperimentResultRow]:
+        processed_count = 0
+        while True:
+            with self._lock:
+                if processed_count < len(self._results):
+                    yield self._results[processed_count]
+                    processed_count += 1
+                elif not self._thread.is_alive():
+                    break
+
+    def _process_data(self, manager: _ExperimentManager) -> None:
+        tqdm = _load_tqdm()
+        results = manager.get_results()
+        for item in tqdm(results):
+            with self._lock:
+                self._results.append(item)
+        batch_scores = manager.get_batch_scores()
+        with self._lock:
+            self._aggregate_results = batch_scores
+
+    def __len__(self) -> int:
+        return len(self._results)
+
+    def __repr__(self) -> str:
+        return f"<ExperimentResults {self.experiment_name}>"
+
+    def wait(self) -> None:
+        self._thread.join()
+
+
+## Private API
+
+
+def _evaluate(
+    target: Union[TARGET_T, Iterable[schemas.Run]],
+    /,
+    data: DATA_T,
+    evaluators: Optional[Sequence[EVALUATOR_T]] = None,
+    summary_evaluators: Optional[Sequence[SUMMARY_EVALUATOR_T]] = None,
+    metadata: Optional[dict] = None,
+    experiment_prefix: Optional[str] = None,
+    max_concurrency: Optional[int] = None,
+    client: Optional[langsmith.Client] = None,
+    blocking: bool = True,
+) -> ExperimentResults:
+    manager = _ExperimentManager(
+        data,
+        client=client,
+        metadata=metadata,
+        experiment_prefix=experiment_prefix,
+        runs=None if callable(target) else target,
+    ).start()
+    if callable(target):
+        manager = manager.with_predictions(target, max_concurrency=max_concurrency)
+    if evaluators:
+        manager = manager.with_scores(evaluators, max_concurrency=max_concurrency)
+    if summary_evaluators:
+        manager = manager.with_batch_scores(summary_evaluators)
+    results = ExperimentResults(manager)
+    if blocking:
+        results.wait()
+    return results
+
+
+def _is_uuid(value: str) -> bool:
+    try:
+        uuid.UUID(value)
+        return True
+    except ValueError:
+        return False
+
+
+def _load_nested_traces(
+    project: Union[str, uuid.UUID], client: langsmith.Client
+) -> List[schemas.Run]:
+    if isinstance(project, uuid.UUID) or _is_uuid(project):
+        runs = client.list_runs(project_id=project)
+    else:
+        runs = client.list_runs(project_name=project)
+
+    treemap: DefaultDict[uuid.UUID, List[schemas.Run]] = collections.defaultdict(list)
+    results = []
+    all_runs = {}
+    for run in runs:
+        if run.parent_run_id is not None:
+            treemap[run.parent_run_id].append(run)
+        else:
+            results.append(run)
+        all_runs[run.id] = run
+    for run_id, child_runs in treemap.items():
+        all_runs[run_id].child_runs = sorted(child_runs, key=lambda r: r.dotted_order)
+    return results
```

Comment:
why don't we just map trace_id to list of runs, then sort each entry based on dotted order?

Reply:
Still need to reconstruct the tree right?

I could also just default to only the roots, which is all people need in 90% of usage rn